components/camera/videosource/webcam.go

```diff
@@ -208,38 +213,89 @@ func findCamera(ctx context.Context, attrs *WebcamAttrs, logger golog.Logger) (c
+	}
+	info, err := gostream.InfoFromMediaSource[image.Image, prop.Video](src)
+	if err != nil {
+		return "", errors.Wrap(err, "could not get info from media source")
+	}
+
+	return getLabelFromDriverInfo(info), nil
+}
+
+
 // NewWebcamSource returns a new source based on a webcam discovered from the given attributes.
 func NewWebcamSource(ctx context.Context, attrs *WebcamAttrs, logger golog.Logger) (camera.Camera, error) {
 	cam, err := findCamera(ctx, attrs, logger)
 	if err != nil {
 		return nil, errors.Wrap(err, "could not find video source for camera")
 	}
 
-	goutils.PanicCapturingGo(func() {
-		src, err := camera.SourceFromCamera(cam)
+	label := attrs.Path
+	if label == "" {
+		label, err = getLabelFromCamera(cam)
 		if err != nil {
-			logger.Debugw("cannot get source from camera", "error", err)
-			return
+			logger.Debugw("cannot find label for camera", "error", err)
+
 		}
+	}
 
-		defer func() {
-			logger.Debug("closing camera")
-			goutils.UncheckedError(cam.Close(ctx))
-		}()
+
+	isConnected := true
+	constraints := makeConstraints(attrs, attrs.Debug, logger)
+	webcamConnectionFn := func() {
 		for {
-			if goutils.SelectContextOrWait(ctx, 500*time.Millisecond) {
-				// mediadevices connects to the OS to get the properties for a driver. If the OS no longer knows
-				// about a specific driver then properties will be empty, and we can safely assume the driver no
-				// longer exists and should be closed.
-				if len(gostream.PropertiesFromMediaSource[image.Image, prop.Video](src)) == 0 {
-					return
+			if !goutils.SelectContextOrWait(ctx, 500*time.Millisecond) {
+				return
+			}
+
+			if isConnected {
+				// If we cannot get the source or driver from our camera then we cannot know if it's connected.
+				// It's better to not assume it is or isn't and retry.
+				src, err := camera.SourceFromCamera(cam)
+				if err != nil {
+					logger.Errorw("cannot get source from camera", "error", err)
+					continue
+				}
+
+				props, err := gostream.PropertiesFromMediaSource[image.Image, prop.Video](src)
+				if err != nil {
+					logger.Errorw("cannot get properties from media source", "error", err)
+					continue
+				}
+
+				// https://github.com/pion/mediadevices connects to the OS to get the properties for a driver. If the
+				// OS no longer knows about a specific driver then properties will be empty, and we can safely assume the
+				// driver no longer exists and should be closed.
+				if len(props) == 0 {
+					isConnected = false
+					logger.Debugf("camera '%v' disconnected", label)
+					goutils.UncheckedError(cam.Close(ctx))
+					continue
 				}
 			} else {
-				return
+				// If we cannot get the source or overwrite the camera's source we try again.
+				src, err := getNamedVideoSource(label, false, constraints, logger)
+				if err != nil {
+					logger.Errorf("could not find video source '%v'", label)
+					continue
+				}
+
+				if err := camera.OverwriteCameraVideoSource(src, cam); err != nil {
```

Comment:
can you use Reconfigure here instead?

Reply:
`Reconfigure` only overrides `actual` and not the `videoSource` and the `videoStream` which depends on the `videoSource`. I could update `Reconfigure` with this logic though if you prefer that.